### PR TITLE
Functions in calc

### PR DESCRIPTION
**Pull request by Bjvanminnen**
I had thought that functions already worked in calc from my last big Calc PR, but it seems that I was wrong. This gets us most of the way there, though there are still a bunch of TODOs in here. I wanted to get what I have committed so that levels like staging.learn.code.org/s/algebra/stage/11/puzzle/6 work again.

image

What in this
- For function calls display in the form foo(1, 2)
  - if the name is wrong i.e. bar(1,2), everything gets marked in red, since the fix is to replace the bar block with the foo block
  - if we have the wrong number of args, i.e. foo(1) or foo(1, 2, 3) we mark everything in parens red. this is maybe a little confusing, since the fix is to actually edit the foo function to take the right number of args. in this case we also want to make sure it's marked in the function signature, but don't support that yet
- For function definitions display in the form foo(x,y) = (x + y)
  - To reiterate the point above, if we instead do foo(x, y, z) = (x + y) we really want to mark the fact that the left side of this is wrong, but won't currently.

### CODE REVIEW

File: apps/src/calc/calc.js [hunk 1]
```diff
@@ -295,22 +296,38 @@ function getEquationFromBlock(block) {
...
       var def = Blockly.Procedures.getDefinition(name, Blockly.mainBlockSpace);
-      if (!def.isVariable()) {
-        throw new Error('not expected');
+      if (def.isVariable()) {
+        return new Equation(null, new ExpressionNode(name));
+      } else {
+        var values = [];
+        var i = 0;
+        var input, childBlock;
+        while (!!(input = block.getInput('ARG' + i))) {
```
Comment (bcjordan) [inline apps/src/calc/calc.js]: Not sure if this is any better:
```
for (var i = 0; !!(input = block.getInput('ARG' + i)); i++) {
}
```

Comment (Bjvanminnen) [inline apps/src/calc/calc.js]: actually, i think i could also just do
```
while (!!(input = block.getInput('ARG' + i++))) {
```
and get rid of the i++ at the bootm

File: apps/src/calc/expressionNode.js [hunk 2]
```diff
@@ -185,24 +185,40 @@ ExpressionNode.prototype.collapse = function () {
...
+      new Token(this.value, this.value !== other.value),
+      new Token('(', !nodesMatch)
+    ];
+
+    for (var i = 0; i < this.children.length; i++) {
+      tokens.push(this.children[i].getTokenListDiff(nodesMatch && other.children[i]));
+      tokens.push(new Token(',', !nodesMatch));
+    }
+    // pull off last comma
+    tokens.splice(-1);
```
Comment (bcjordan) [inline apps/src/calc/expressionNode.js]: `this.children.length` is always > 0?

Comment (Bjvanminnen) [inline apps/src/calc/expressionNode.js]: Nope. good catch. looks like a good test case too :)

**Comment by bcjordan** [comment]
LGTM, except for the no children case. Love the test setup

**Comment by Bjvanminnen** [comment]
Upon looking into it more, I realized a function call can't have no children. If it does, it's really a variable. I added support and a test case for that, and refactored that code some. If you could take another look at it, that would be cool.

I also changed the loop to your suggestion.

**Pull request merged**